torch/_masked/__init__.py

```diff
@@ -390,21 +396,71 @@ def _canonical_dim(dim: DimOrDims, ndim: int) -> Tuple[int, ...]:
+    with the same shape as input and with (broadcasted) content of
+    mask, if specified.
+
+    For sparse COO inputs, the canonical input mask is a sparse
+    coalesced tensor with the same shape and indices as the coalesced
+    :attr:`input` tensor.  The values of the canonical input masks are
+    all True, unless the :attr:`mask` tensor dtype is torch.bool, in
+    which case the :attr:`mask` values define the canonical input
+    mask.
+
+    If specified :attr:`mask` tensor uses a different layout from the
+    :attr:`input` tensor, then the canonical input mask tensor will
+    use the layout of the :attr:`input` tensor.
+
     """
+    if input.layout not in {torch.strided, torch.sparse_coo}:
+        raise ValueError(f'_input_mask expects strided or sparse COO tensor but got {input.layout}')
+
     mask = kwargs.get('mask')
+
+    # default mask
     if mask is None:
-        inmask = input.new_ones(input.shape, dtype=torch.bool)
-    elif mask.ndim < input.ndim:
-        inmask = torch.broadcast_to(mask.clone(), input.shape).to(dtype=torch.bool)
-    elif mask.ndim > input.ndim:
-        raise IndexError("_input_mask expected broadcastable mask (got mask dimensionality higher than of the input)")
-    elif mask.shape != input.shape:
-        inmask = torch.broadcast_to(mask.clone(), input.shape).to(dtype=torch.bool)
+        if input.layout == torch.strided:
+            mask = input.new_ones(input.shape, dtype=torch.bool)
+        else:
+            mask = input
+
+    # mask shape must match with input shape
+    if mask.shape != input.shape:
+        if mask.ndim > input.ndim:
+            raise IndexError("_input_mask expected broadcastable mask (got mask dimensionality higher than of the input)")
+        if mask.layout == torch.strided:
+            mask = torch.broadcast_to(mask.clone(), input.shape).to(dtype=torch.bool)
+        else:
+            mask = torch._sparse_broadcast_to(mask, input.shape)
+
+    # mask layout must match with input layout
+    if mask.layout != input.layout:
+        if input.layout == torch.strided:
+            if mask.dtype == torch.bool and not mask.is_coalesced():
+                # to_dense calls coalesce but coalesce not implemented for Bool
+                mask = mask.to(torch.uint8).coalesce().to(torch.bool)
+            mask = mask.to_dense()
+        else:
+            mask = mask.to_sparse(input.sparse_dim())
+
+    # mask is a boolean tensor
+    if input.layout == torch.strided:
+        if mask.dtype != torch.bool:
+            mask = mask.to(dtype=torch.bool)
     else:
-        inmask = mask.to(dtype=torch.bool)
-    return inmask
+        if mask.dtype != torch.bool:
+            if not mask.is_coalesced():
+                mask = mask.coalesce()
+            values = mask.values()
+            mask = torch.sparse_coo_tensor(mask.indices(),
+                                           values.new_ones(values.shape, dtype=torch.bool),
```

Comment:
I think instead we could convert to torch.bool and coalesce to remove materialized False entries.

Reply:
This note should apply to the "dtype is bool" case. Consider the following usage:
```
torch._masked.op(input, mask=None)
```
that is currently equivalent to (Code 1):
```
torch._masked.op(input, mask=input.new_ones(input.shape, dtype=bool) if input.layout == torch.strided else input)
```
that is, in the case of sparse input, its sparsity pattern determines the mask and input values are ignored. From this follows that for sparse input as in
```
torch._masked.op(input, mask=input)
```
the mask always ignores input values.

Hmm, should we change Code 1 to
```
torch._masked.op(input,
  mask=(input.new_ones(input.shape, dtype=bool) if input.layout == torch.strided
        else input.to(dtype=bool).coalesce()))
```
?

That is, the canonical mask will be [showing definitions, actual implementations may differ]
- `mask.to(dtype=bool).coalesce()` for COO tensor
- `mask.to_sparse().to(dtype=bool).coalesce().to_sparse_csr()` for CSR tensor
- `mask.to(dtype=bool)` for strided tensor
Would that make more sense?

Reply:
Yes, I think your proposal makes more sense, but also I'd want to avoid diversion from the strided case (since we want ops to commute with to_dense). How about if mask is None the op simply applies the regular reduction?

For strided input we'd simply map `torch._masked.reduction_op -> torch.reduction_op`.

For sparse input we could potentially implement it as (here also just a definition)
```
torch._masked.reduction_op(input, mask):
  if mask is None:
    # zeros are masked for performance.
    result = torch._masked.reduction_op(input, mask=input.to(bool))
    return pointwise_binary_op(result, torch.tensor(0)) # sparse output
```
This behavior should cause `torch._masked.reduction_op(input) <-> torch.reduction_op(input)` for sparse as well
This would satisfy the use case where users want to respect the zeros and maintain our to_dense constraint.
If a user wants to mask the zeros they'd just call it with `mask=input.to(bool)` or so.

Reply:
@cpuhrsch , yes, when `mask is None`, it makes sense to fall back to regular reduction. The only problem here is that we have not defined yet what is the regular reduction on sparse tensors in general. The proposed definition above perfectly applies to `sum` because the unspecified values can be conveniently interpreted as zeros. But we have `prod`, `min`, 'max', etc where we need to specify how the unspecified values are interpreted as we don't have regular reductions implemented on sparse inputs yet (sum is an exception here).

For the record, interpreting the unspecified values as zeros, in general, does not sound particularly useful. For instance, `torch._masked.prod` on a sparse tensor would result most likely in a zero tensor unless sparsity is close to zero. Another example is `min/max` on sparse tensors that would lead to sparsity patterns that correspond to "is negative/positive" tests.

Your definition suggests: if `mask is None` then define `mask = input.to(bool)`. While this can be considered as the definition of regular reduction for any input layout, it is biased to interpreting explicit zeros as masked-out values. In fact, from this definition follows that the result of regular reduction on any input is never zero if the reduction operations is `prod`, or `min`, or `max`, etc. Hence, `torch._masked.reduction_op(input) != torch.reduction_op(input)`, in general (including strided inputs, exceptions are `sum` and the cases with inputs that do not contain any zero values).

So, while being back to square one, let me review what we have and what we need to define:
1. For strided inputs, we define `torch._masked.reduction_op(input) == torch.reduction_op(input)`. This basically means that the default mask (when specified as None) is `mask = input.new_ones(input.shape, dtype=bool)` for all reduction operations.
2. For sparse inputs, we don't have explicit definitions nor implementations for `torch.reduction_op(input)` (the exception is sum in terms of torch.sparse.sum but this is not sufficient), so it does not make sense to refer to `torch.reduction_op` behaviour when talking about sparse inputs.

However, within masked semantics (that `torch._masked` is all about), we know that the default mask 
(i) would have the same layout as the input (to minimize the need to implement mixed-layout binary operations),
(ii) it should not depend on the values in the input (otherwise the sparsity pattern of the reduction operation result will be variable for the fixed sparsity pattern of the input with variable values),
(iii) the strided input case is well defined as explained in point 1. above,
and (iv), in the sparse input cases, the sparsity pattern of the default mask cannot have fewer unspecified values than the number of unspecified values in the input, that is, `mask.values().numel() <= input.values().numel()` must always hold.

I think the most natural default mask for sparse inputs that satisfies the above criteria would be:
```
mask = torch.<input.layout>_tensor(<input indices>, input.values().new_ones(input.values().shape), dtype=bool)
```
Notice that this is not equivalent to `mask=input.to(dtype=bool)` that violates (ii).

@cpuhrsch what do you think?

Reply:
I think one fundamental assumption thus far has been that we want the behavior of functions to not depend on the layout of the inputs. So whether the input is sparse or dense, the output should be the same. If the mask is None, the masked reduction is equivalent to a regular reduction or, as you point out, a mask full of `True`. That means there is no default mask, the user has to specify one. Then, if the mask is not None, the same applies and we can orient ourselves on the existing implementation for the strided case to specify the behavior.

Now, from an implementation perspective, if it is currently hard or very inefficient to support a certain combination of inputs (say strided mask and sparse input (example chosen for no particular reason)) let us throw an error and incrementally add functionality based on user feedback.

Reply:
@cpuhrsch there is one important property that I didn't take into account: by default, the sparse tensors represent compressions of zeros, that is, the unspecified elements are treated as zero elements. This defines `torch.reduction_op` on sparse tensors, and as a result, we have the definition for `torch._masked.reduction_op` with `mask is None` case.
This also satisfies the fundamental assumption about layout invariance of functions behavior.

I think I now have all the needed definitions to review the current implementation accordingly and perhaps document the fundamental principles of the implementation choices we have discussed here.